src/evict.c

```diff
@@ -329,18 +329,23 @@ unsigned long LFUDecrAndReturn(robj *o) {
  * returns the sum of AOF and slaves buffer. */
 size_t freeMemoryGetNotCountedMemory(void) {
     size_t overhead = 0;
-    int slaves = listLength(server.slaves);
-
-    if (slaves) {
-        listIter li;
-        listNode *ln;
 
-        listRewind(server.slaves,&li);
-        while((ln = listNext(&li))) {
-            client *slave = listNodeValue(ln);
-            overhead += getClientOutputBufferMemoryUsage(slave);
+    /* Since all replicas and replication backlog share global replication
+     * buffer, we think only the part of exceeding backlog size is the extra
+     * separate consumption of replicas. */
```

Comment:
maybe that would be ok, but only if we don't run into an endless feedback loop on mass eviction, which causes over-eviction and data loss.

so i'd like to imagine 3 cases.
let's say we have a server with maxmemory=1gb, and the backlog limit set to 100mb.
1. if the backlog is full and the replicas are not lagging, eviction deletes keys and the backlog normally remains 100mb, so other than a small spike because a new repl-buffer node is created it shouldn't cause an infinite feedback loop.
2. if there's a lagging replica that causes the repl buffers to grow to 500mb, each eviction would cause the buffer to grow, and had we counted that we could get into a feedback loop. but since we ignore the part above the first 100mb, we're ok.
3. if there are no replicas connected for a long time, so also no backlog, and the db is full constantly doing evictions. then a replica connects and the backlog and repl buffers start to grow, the eviction causes it to grow even faster, and it's now consuming 10mb. each (small) key being deleted inflates the memory further and the eviction loop keeps running until the backlog reaches 100mb, and which point the feedback loop stops and eviction eventually completes.

let's analyse case 3.
- since we now have incremental eviction, we don't have the risk of staying in an infinite loop, we'll re-sample the memory state when we re-enter the performEvictions function.
- since (unlike the old slave buffers), the backlog will keep consuming that memory even after the eviction loop stops and the replicas consume their buffers, we won't go into a resonance effect. 

the benefit of keeping the current proposal that doesn't ignore the backlog part (first 100mb in our case), is that on upgrades from previous versions, users won't suddenly be able to push another 100mb before reaching eviction.

so bottom line, i'm changing my mind and i'm ok with this.
still, i'd like to turn the attention of others to this discussion.
let's mention it in the top comment and hope they'll read it when the time comes.

Reply:
1. 
```
AFAIK the reason the replica buffers and aof buffers are not counted for eviction was because it can cause feedback-loop when we push DELs into them.
```
I think this is tribal knowledge that is not documented anywhere in the code. This is the first time I'm learning this. Could we please document it?

2. I agree with the current logic.

Reply:
yes, let's document that, and also big chunk of my previous comment, i.e. describe why not counting the backlog size would not cause a feedback loop concern.

Reply:
Here, i don't judge `listLength(server.slaves)`
`if ((long long)server.repl_buffer_mem > server.repl_backlog_size) `
i.e. i still regard the part of exceeding backlog size as the consumption of replica because we don't trim replication backlog in time, to avoid eviction.

@oranagra do you think it make sense?

Reply:
yes, it makes sense to me..
specifically because the only reason there would be a very long list of excess nodes is when a replica dropped.
so had we counted the memory that's actually being used by replicas, as long as they're connected we'd ignore that memory, and as soon as they drop we run into a massive eviction loop.
let's document that too (i.e. that because the deletion is in the background, and has a delayed effect after a replica disconnects, we have to ignore it even if there are no replicas)